src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectSnapshot.cs

```diff
@@ -58,25 +48,36 @@ public ProjectSnapshot(ProjectState state)
 
     public ProjectWorkspaceState ProjectWorkspaceState => State.ProjectWorkspaceState;
 
-    public virtual IDocumentSnapshot? GetDocument(string filePath)
+    public bool ContainsDocument(string filePath)
     {
-        lock (_lock)
+        lock (_gate)
         {
-            if (!_documents.TryGetValue(filePath, out var result) &&
-                State.Documents.TryGetValue(filePath, out var state))
-            {
-                result = new DocumentSnapshot(this, state);
-                _documents.Add(filePath, result);
-            }
-
-            return result;
+            return _filePathToDocumentMap.ContainsKey(filePath) ||
+                   State.Documents.ContainsKey(filePath);
         }
     }
 
+    public IDocumentSnapshot? GetDocument(string filePath)
+        => TryGetDocument(filePath, out var document)
+            ? document
+            : null;
+
     public bool TryGetDocument(string filePath, [NotNullWhen(true)] out IDocumentSnapshot? document)
     {
-        document = GetDocument(filePath);
-        return document is not null;
+        lock (_gate)
+        {
+            // We only create a new DocumentSnapshot if we haven't created one yet
+            // but have DocumentState for it.
+            if (!_filePathToDocumentMap.TryGetValue(filePath, out var snapshot) &&
+                State.Documents.TryGetValue(filePath, out var state))
```

Comment:
Nit: This little double-condition has annoyed me before. Seems like now might be a good opportunity to check `_filePathToDocumentMap` separately and return early, and make this thing more readable.

Reply:
Sure thing!